### PR TITLE
ARROW-5977: [C++] [Python] Allow specifying which columns to include

### PR DESCRIPTION
**Pull request by pitrou**
If the `include_columns` option is not empty:
- columns are included in that order
- columns not in `include_columns` are ignored
- columns in `include_columns` but not in the CSV file produce a null column

### CODE REVIEW

File: python/pyarrow/tests/test_csv.py [hunk 1]
```diff
@@ -306,6 +312,58 @@ def test_header_column_names(self):
             "y": ["kl", "op"],
             }
 
+    def test_include_columns(self):
```
Comment (pitrou) [inline python/pyarrow/tests/test_csv.py]: @nealrichardson You may want to look at the tests below and validate them.

File: python/pyarrow/tests/test_csv.py [hunk 2]
```diff
@@ -306,6 +312,58 @@ def test_header_column_names(self):
...
+            }
+
+        # Order of include_columns is respected, regardless of CSV order
+        convert_options.include_columns = ['cd', 'ab']
+        table = self.read_bytes(rows, read_options=read_options,
+                                convert_options=convert_options)
+        schema = pa.schema([('cd', pa.string()),
+                            ('ab', pa.string())])
+        assert table.schema == schema
+        assert table.to_pydict() == {
```
Comment (nealrichardson) [inline python/pyarrow/tests/test_csv.py]: Is this an `OrderedDict`?

File: python/pyarrow/tests/test_csv.py (hunk 1 above)
Comment (nealrichardson) [reply python/pyarrow/tests/test_csv.py]: LGTM aside from my one question about order

File: python/pyarrow/tests/test_csv.py (hunk 2 above)
Comment (pitrou) [reply python/pyarrow/tests/test_csv.py]: No, but the order is testing using schema equality above.

**Comment by wesm** [comment]
On this item

> columns in include_columns but not in the CSV file produce a null column

I'm not certain if this is a bridge too far -- should the introduction of null columns be handled elsewhere (genuine question, I'm not sure)?

**Comment by nealrichardson** [comment]
I could see having that case error, but was thinking that it would be useful when reading a multi-file dataset where you have a schema for the dataset but not all files have all columns, so you need to fill in with missings for that chunk. 

At the risk of adding complexity, could there be some flag or setting that governs whether specifying a nonexistent column to include errors or creates a null column? Or some way of signaling a warning that null columns were created?

**Comment by pitrou** [comment]
> At the risk of adding complexity, could there be some flag or setting that governs whether specifying a nonexistent column to include errors or creates a null column?

There could... The question is: is it useful?

**Comment by jorisvandenbossche** [comment]
From a user perspective for reading a single file, I think having it error on unknown column names is useful behaviour to avoid surprises. 

> would be useful when reading a multi-file dataset where you have a schema for the dataset

Could this option be provided by the ability to specify the full schema you want? And in that case create all-null columns (but with given type) if a column is present in the schema but not in the file.

**Comment by pitrou** [comment]
Hmm, so based on the last messages, it will be either:
* raise on missing column
* add an option to turn a missing column into a null column (false by default, meaning raise)

**Comment by pitrou** [comment]
I added a `include_missing_columns` option. Please advise.

**Comment by wesm** [comment]
SGTM. @nealrichardson or @jorisvandenbossche, any thoughts?

**Comment by nealrichardson** [comment]
I think it's ok. I'm generally hesitant to start adding lots of arguments to function signatures, but file readers tend to grow lots of them for good reason. `include_missing_columns` could be YAGNI I guess, but I don't feel strongly enough about that to advocate removing the option (and having the behavior be to error if you specify cols that don't exist, as Joris proposed). But I would support you if you thought it was YAGNI and wanted to remove it until it was needed.

**Comment by jorisvandenbossche** [comment]
Fine for me as well. 

Although if we add a `include_missing_columns` keyword, I wonder how it should interact with eg the `column_types` option. I would expect that it has effect on that as well, if a non-existent column is specified there.

BTW, if we start having an option for how to handle "missing" columns, there is actually also a third possibility: ignore them (no error, but also not include them as null columns). So instead of a boolean option, it would also have 3 string options (although that might even be more over the top).

> file readers tend to grow lots of them for good reason

Indeed, look at `pandas.read_csv` ... It would be nice to somehow not repeat that, but in the end you often need quite some options to work with messy real-world data.

**Comment by jorisvandenbossche** [comment]
@pitrou how does the `include_columns` + `include_missing_columns=True` interact with `column_types` ? Does it still give a null type or does it actually use the type specified in `column_types` ?

**Comment by pitrou** [comment]
@jorisvandenbossche It still gives a null type. Do you think that needs to be changed? I'm starting to think that `ignore_missing_columns` may not be desirable after all...

**Comment by jorisvandenbossche** [comment]
> Do you think that needs to be changed?

I would personally expect it to give the type as specified in `column_types`. But it is certainly a rather complex interaction between the different keywords.

However, for such a usecase, you already need to specify a combination of `column_types`, `include_columns` and `include_missing_columns=True`, which might be easier with "just" a way to specify the final desired schema of the table (then there is a single keyword that can be used for this use case).

**Comment by pitrou** [comment]
The thing is, `column_types` allows you to force-type certain columns while inferring others... Passing the entire schema isn't as flexible.

**Comment by nealrichardson** [comment]
The reason we were considering adding null columns was so that you could essentially align the schema of multiple CSVs that may not all have the same columns. You would basically use the CSV reader arguments to define the schema you want to have and then it would fit the file to it as best as it could. 

Maybe that's better handled outside of the CSV reader--not worth the added complexity?

**Comment by pitrou** [comment]
I'm not sure how you would handle it outside of the CSV reader, though, since the reader would just error out.

**Comment by jorisvandenbossche** [comment]
> I'm not sure how you would handle it outside of the CSV reader, though, since the reader would just error out.

We will need a schema normalization / evolution functionality for the datasets work anyhow I think? (or will that be done by each reader separately) In this case it could mean adding additional (null) columns to the table after it was read by the csv reader.

**Comment by pitrou** [comment]
Perhaps... but at least if filtering existing columns, it's more efficient to filter them out at the CSV reader level, than to convert them in full (consuming CPU cycles and memory) then drop them in a subsequent schema normalization step.

**Comment by jorisvandenbossche** [comment]
Although, the datasets layer above CSV would then need to know which columns are in the csv to not ask the csv file to read missing columns, so that might indeed be a bit difficult.

I think, in the end, your current solution with a toggle to allow missing columns is quite OK.

**Comment by pitrou** [comment]
I can try to implement the interaction with `column_types` tomorrow anyway. Shouldn't be _that_ difficult...

**Comment by nealrichardson** [comment]
Filtering out in the CSV reader is efficient and is the objective here; we're debating whether to use the CSV reader to add columns that don't exist or whether that should happen outside of the reader, IIUC.

**Comment by pitrou** [comment]
@nealrichardson, see @jorisvandenbossche observation:
> Although, the datasets layer above CSV would then need to know which columns are in the csv to not ask the csv file to read missing columns, so that might indeed be a bit difficult.

**Comment by fsaintjacques** [comment]
As @jorisvandenbossche said, we will need a reader that takes a Schema as input. The Schema would dictate the following:

- Column selection
- Type without inference

This could be done in a separate csv::Reader. I don't think we should mix this with the current setting (yet).

**Comment by pitrou** [comment]
Ok, I implemented the interaction between `column_types` and `include_missing_columns`.

**Comment by jorisvandenbossche** [comment]
Looks good to me now!

**Comment by pitrou** [comment]
Ok, will merge.

**Pull request closed without merging**